File: 146 (LRU Cache)/cpp/Solution.cpp
class DoubleNode{
    public:
    int key, val;
    DoubleNode* next;
    DoubleNode* prev;
    DoubleNode(int key, int val){
        this->key = key;
        this->val = val;
        prev= nullptr;
        next = nullptr;
    }
};


class LRUCache {
public:
    int capacity;
    unordered_map<int, DoubleNode*> dic;
    DoubleNode* left;
    DoubleNode* right;
    LRUCache(int capacity) {
        this->capacity = capacity;
        left = new DoubleNode(-1,-1);
        right = new DoubleNode(-1,-1);
        left->next = right;
        right->prev = left;
    }
    
    int get(int key) {
        if (dic.find(key)!=dic.end()){
            remove(dic[key]);
            insert(dic[key]);
            return dic[key]->val;
        }
        return -1;
    }

    
    void put(int key, int value) {
        if (dic.find(key)!= dic.end()){
            remove(dic[key]);
        }
        dic[key] = new DoubleNode(key,value);
        insert(dic[key]);

        if (dic.size()>capacity){
            DoubleNode* lru = left->next;
            remove(lru);
            dic.erase(lru->key);
        }
    }

    void remove(DoubleNode* node){
        DoubleNode* prevnode = node->prev;
        DoubleNode* nextnode = node->next;
        prevnode->next = nextnode;
        nextnode->prev = prevnode;
    }

    void insert(DoubleNode* node){
        DoubleNode* prevnode = right->prev;
        DoubleNode* nextnode=  right;
        prevnode->next = node;
        node->prev=  prevnode;
        node->next = nextnode;
        nextnode->prev = node;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 146 (LRU Cache)/python/Solution.py
class ListNode:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.next = None
        self.prev= None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.left = ListNode(-1,-1)
        self.right = ListNode(-1,-1)
        self.left.next = self.right
        self.right.prev= self.left
        self.dic = {}

    def get(self, key: int) -> int:
        if key in self.dic:
            self.remove(self.dic[key])
            self.insert(self.dic[key])
            return self.dic[key].val
        
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.dic:
            self.remove(self.dic[key])
        self.dic[key] = ListNode(key,value)
        self.insert(self.dic[key])

        if len(self.dic)>self.capacity:
            lru = self.left.next
            self.remove(lru)
            del self.dic[lru.key]

    def insert(self,node):
        prevnode, nextnode = self.right.prev, self.right
        prevnode.next = node
        node.prev = prevnode
        node.next = nextnode
        nextnode.prev= node
    
    def remove(self,node):
        prevnode, nextnode = node.prev, node.next
        prevnode.next = nextnode
        nextnode.prev = prevnode
    

        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)